source/CScriptEngine.cpp

```diff
@@ -131,15 +129,115 @@ namespace CLEO
         return (SCRIPT_VAR*)((size_t)result + pScript->GetBasePointer());
     }
 
-    void __fastcall _GetScriptStringParam(CRunningScript *pScript, int dummy, char *buf, int len)
+    char* __fastcall GetScriptStringParam(CRunningScript* thread, int dummy, char* buff, int buffLen)
     {
-        _asm
+        if (buff == nullptr || buffLen == 0)
+            return buff;
+
+        if (buffLen > 0)
+            ZeroMemory(buff, buffLen);
+        else
+            buffLen = 0x7FFFFFFF; // unknown - unlimited
+
+        auto paramType = thread->PeekDataType();
+        auto arrayType = IsArray(paramType) ? thread->PeekArrayDataType() : eArrayDataType::ADT_NONE;
+
+        auto isVariableInt = IsVariable(paramType) && (arrayType == eArrayDataType::ADT_NONE || arrayType == eArrayDataType::ADT_INT);
+
+        // integer address to text buffer
+        if (IsImmInteger(paramType) || isVariableInt)
         {
-            mov ecx, pScript
-            push len
-            push buf
-            call FUNC_GetScriptStringParam
+            GetScriptParams(thread, 1);
+
+            if (opcodeParams[0].dwParam <= CCustomOpcodeSystem::MinValidAddress)
+            {
+                LOG_WARNING(thread, "Invalid '0x%X' pointer of input string argument #%d in script %s", opcodeParams[0].dwParam, CLEO_GetParamsHandledCount(), ScriptInfoStr(thread).c_str());
+                return nullptr; // error
+            }
+
+            auto len = min((int)strlen(opcodeParams[0].pcParam), buffLen);
+            memcpy(buff, opcodeParams[0].pcParam, len);
+            if (len < buffLen) buff[len] = '\0'; // add terminator if possible
+            return buff;
+        }
+        else if (paramType == DT_VARLEN_STRING)
+        {
+            thread->IncPtr(1); // already processed paramType
+
+            DWORD length = *thread->GetBytePointer(); // as unsigned byte!
+            thread->IncPtr(1); // length info
+
+            char* str = (char*)thread->GetBytePointer();
+            thread->IncPtr(length); // text data
+
+            auto len = min((int)length, buffLen);
+            memcpy(buff, str, len);
+            if (len < buffLen) buff[len] = '\0'; // add terminator if possible
+            return buff;
         }
+        else if (IsImmString(paramType))
+        {
+            thread->IncPtr(1); // already processed paramType
+
+            auto str = (char*)thread->GetBytePointer();
+
+            switch (paramType)
+            {
+                case DT_TEXTLABEL:
+                {
+                    auto len = min((int)strnlen(str, 8), buffLen);
+                    memcpy(buff, str, len);
+                    if (len < buffLen) buff[len] = '\0'; // add terminator if possible

```

Comment:
Yes, but memcpy copies just the text without terminator. 
Provided target buffer might be different size (greater than source) so checks are needed anyway. Maybe it can be slightly rearranged for this two cases, but seems better to just keep same logic everywhere.

Reply:
you don't need to calculate the length either. thing is, both DT_TEXTLABEL and DT_STRING are static byte sequences, 8 or 16 chars long. you just need to copy them as-is into the destination buffer. somewhere in the sequence they will have a null-terminator as well. it will be simple and efficient was of reading these values.

```
 case DT_TEXTLABEL:
{
memcpy(buff, str, 8);
thread->IncPtr(8); // text data
return buff;
}
```

Reply:
we need to keep in mind, that executing a bytecode should be fast enough to handle thousands of opcodes per frame. adding more and more validations is good, but reduces the performance. some of these checks were intentionally implemented on script side, not in runtime specifically to increase the speed of work,

Reply:
Updated.